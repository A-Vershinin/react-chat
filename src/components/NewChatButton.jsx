import React, { Component, Fragment } from 'react';
import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import AddIcon from '@material-ui/icons/Add';
import Modal from '@material-ui/core/Modal';
import Typography from '@material-ui/core/Typography';
import Paper from '@material-ui/core/Paper';
import TextField from '@material-ui/core/TextField';
import blue from '@material-ui/core/colors/blue';

const styles = theme => ({
  newChatButton: {
    position: 'absolute!important',
    left: 'auto',
    right: theme.spacing.unit * 3,
    bottom: theme.spacing.unit * 3 + 48,
  },
  modalWrapper: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  modal: {
    outlineColor: blue[900],
    width: '40%',
    minWidth: '400px',
    minHeight: '250px',
    padding: theme.spacing.unit * 3
  }
});

class NewChatButton extends Component {
  state = {
    open: false,
    title: {
      value: '',
      isValid: true,
    }
  }

  toggleModal = () => {
    this.setState({ open: !this.state.open })
  }

  handleTitleChange = (event) => {
    this.setState({
      title: {
        value: event.target.value,
        isValid: true,
      }
    });
  }

  handleCreateChat = (event) => {
    event.preventDefault();
    const { title } = this.state;

    if (!title.value) {
      this.setState({
        title: {
          value: title.value,
          isValid: false
        }
      })
      return;
    }

    this.props.onCreateChat(title.value);
    this.toggleModal();
    this.setState({
      title: {
        value: '',
        isValid: true
      }
    });
  }

  render() {
    const { open, title } = this.state;
    const { classes } = this.props;

    return (
      <Fragment>
        <Button variant="fab" color="primary"
          className={classes.newChatButton}
          onClick={this.toggleModal}>
          <AddIcon />
        </Button>
        <Modal
          open={open}
          className={classes.modalWrapper}
          onClose={this.toggleModal}
        >
          <Paper className={classes.modal}>
            <Typography variant="title" id="modal-title">
              Create new chat
            </Typography>
            <TextField
              required
              fullWidth
              label="New chat"
              placeholder="Type the title..."
              type="text"
              margin="normal"
              autoComplete="new-chat"
              value={title.value}
              onChange={this.handleTitleChange}
              error={!title.isValid}
            />
            <Button
              color="primary"
              onClick={this.handleCreateChat}
            >
              Create
            </Button>
          </Paper>
        </Modal>
      </Fragment>
    );
  }
}

export default withStyles(styles)(NewChatButton);
