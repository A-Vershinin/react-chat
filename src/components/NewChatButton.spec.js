/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import NewChatButton from './NewChatButton.jsx';

const mockProps = {
  classes: {
    newChatButton: 'some text a',
    modalWrapper: 'some text b',
    modal: 'some text c',
  },
  onCreateChat: jest.fn(),
  disabled: false,
};

describe('<NewChatButton />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <NewChatButton {...mockProps} />
      </MemoryRouter>,
      div,
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<MemoryRouter>
          <NewChatButton {...mockProps} />
              </MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders when NewChatButton disabled', () => {
    const tree = renderer
      .create(<MemoryRouter>
          <NewChatButton {...mockProps} disabled />
              </MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
