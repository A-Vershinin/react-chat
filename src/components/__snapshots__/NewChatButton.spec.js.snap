// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewChatButton /> renders correctly 1`] = `
<button
  className="MuiButtonBase-root-69 MuiButton-root-47 MuiButton-contained-56 MuiButton-fab-64 MuiButton-containedPrimary-57 MuiButton-raised-59 MuiButton-raisedPrimary-60 NewChatButton-newChatButton-44 some text a"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiButton-label-48"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-72"
      color={undefined}
      focusable="false"
      viewBox="0 0 24 24"
    >
      <g>
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </g>
    </svg>
  </span>
  <span
    className="MuiTouchRipple-root-78"
  />
</button>
`;

exports[`<NewChatButton /> renders when NewChatButton disabled 1`] = `
<button
  className="MuiButtonBase-root-69 MuiButtonBase-disabled-70 MuiButton-root-47 MuiButton-contained-56 MuiButton-fab-64 MuiButton-containedPrimary-57 MuiButton-raised-59 MuiButton-raisedPrimary-60 MuiButton-disabled-63 NewChatButton-newChatButton-44 some text a"
  disabled={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="-1"
  type="button"
>
  <span
    className="MuiButton-label-48"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-72"
      color={undefined}
      focusable="false"
      viewBox="0 0 24 24"
    >
      <g>
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </g>
    </svg>
  </span>
</button>
`;
